### DISCUSSION TITLE
suggestions

### DISCUSSION THREAD

**Discussion by wibi9424** [Zergatul/cheatutils#107] [Ideas]
share your suggestions or ideas here

**Comment by wibi9424** [comment]
### another keybinding

so many keybind in my minecraft now, from another mod and this mod. i wonder if i can bind it at chat

example

`/cu toggle esp`
`#toggle esp`
`.tp`

and wonder if i can write script to count some basic math

`/count 21 + 32 - 21`
`.c 5 + 5 * 5 - 5`

**Reply by Zergatul** [reply]
you can only enable "exec" module under scripting. and run things from chat like this:
- `. esp.toggle();`
- `. ui.overlayMessage((21 + 32 - 21).toString());`

**Reply by wibi9424** [reply]
ive try that is awesome, and try many possible script.

umm for long-term maybe typing the same things is not efficient.

**Reply by Zergatul** [reply]
you can also use input API, for example `isKeyDown`, and code your own keys or combinations inside `events.onHandleKeybindings`

**Comment by someRandomDude-a** [comment]
Is it possible to add an auto tool to kill aura which can be configured?

**Reply by Zergatul** [reply]
currently it is not possible. you only want to auto equip single specific item when kill aura attacks something?

**Reply by someRandomDude-a** [reply]
Yep, I want it to auto switch to a weapon when attacking.

Currently I have it setup with baritone so it types out #pause when in range of mobs and holds up a shield, facing the closest creeper, skeleton,zombie while attacking, however, it often ends up punching them with blocks or pickaxe which is undesirable.

Is it possible to make it auto switch to a weapon?

Another case ive been trying to do is auto block any incomming arrows, however its difficult to compute the velocity to figure out when to block. Is there any way to add a more fleshed out system to detect enemies/entities?? 

Is it possible to allow scripting to change esp settings? like changing colors of the glow / overlay??

**Reply by someRandomDude-a** [reply]
But for most cases, yes its going to be a single sword that is used

**Reply by someRandomDude-a** [reply]
Ive also run into roadblocks where the lack of hotkeying is a problem, like for example for a sentry bot, I want to be able to take a screenshot when a player is detected with the freelook camera looking at said player. However I could not figure out how to do it.

Another thing Ive been trying is getting sounds to work so its easy to configure background tasks/alerts can you also explain how to play sounds?

**Reply by Zergatul** [reply]
Adding ability to change colors from scripts is simple. I would never think someone needs that
For screenshots it should be better if I just add new method for scripting. Emulating key is a bit hacky solution.
Even if you want to screenshot players with freecam it will be very hard to position camera properly. You may just stuck camera inside blocks. If you want to screenshot without freecam, you can use method `player.lookAt`
For auto equip I don't know what should be the best way to implement it. I can add new method to equip an item. And you would call it when you stop baritone.
For sounds you have this: https://github.com/Zergatul/cheatutils/blob/1.21.4/common/java/com/zergatul/cheatutils/scripting/modules/SoundApi.java

**Reply by Zergatul** [reply]
You call it like this:
```c#
sound.play("C:\\Users\\Zergatul\\Downloads\\mixkit-bonus-extra-in-a-video-game-2064.ogg");
```

**Reply by Zergatul** [reply]
I added new API methods in 1.21.4 - to set colors, to make screenshot, and to equip item
You can download jar from build here

**Reply by someRandomDude-a** [reply]
The reason for colors is so I can highlight teammates in green and other players in red.

If multiple.players are nearby, the find closest method would not work so is it possible to make an ordered array with the entity IDs based on their distance.

Thankyou so much, the new methods work great

**Reply by Zergatul** [reply]
> The reason for colors is so I can highlight teammates in green and other players in red.

You can't do this
You can only disable some parts of ESP (for example outline) for teammates

**Comment by wibi9424** [comment]
just wondering why in pojav like this
![IMG_20250113_183335](https://github.com/user-attachments/assets/c10fdc04-6914-41c7-bbc9-974961ec70e1)

graphic changing after v 3.0.1 for minecraft fabric 1.21.1, i cant using third person, and hands no texture

just wondering,

.

by the way actually I just want to ask about this

player.target.getBlockX()
and for Y and Z

how do you get the coordinates of the furthest block visible on the screen?

**Reply by Zergatul** [reply]
probably something is messed up with framebuffers and texture ids. what is pojav
why do you need the furthest block visible on the screen? it must be very costly to calculate such thing

**Reply by wibi9424** [reply]
because im experiment in creative mode, that i will summon lightning bolt in the block that i see,

just like `world edit` using compas that when we press right click then we will teleport to that block, there's max range that we cant teleport to that block if that so far away

**Reply by Zergatul** [reply]
ok, so you don't really want "the furthest block visible on the screen", you want to extend "pick" range. because minecraft uses 20 blocks by default

**Reply by Zergatul** [reply]
Added this for 1.21.4, but then realized you are on 1.21.1:
```c#
let pos = player.target.getBlockPosInRange(1000, false); // get block collision with line of sight in 1000 blocks range
if (pos.x != int.MIN_VALUE) {
    // do something
}
```

**Reply by wibi9424** [reply]
yeah, 1.21.4, its because graphics errors in your newest version, maybe caused by sodium or something in pojav, so i stay in 1.21.1,

i know when bugs happen the first problem might be pojav is not compatible or else, but i like to play in pojav than in my laptop, so maybe i just open 1.21.4 to see what's new then back to pojav

**Reply by Zergatul** [reply]
So pojav allows to run minecraft on ios and android? I don't know this was possible before. And I have no idea what can be wrong

**Reply by wibi9424** [reply]
yeah to day is possible, also many shader that can run on android, ive try using optifine on 1.21.4

**Comment by EllieDotP** [comment]
Do any of the other versions work for 1.21.1 neoforge? If not can you please make a version that works for it!

**Reply by Zergatul** [reply]
1.21 should work

**Comment by Dispnser** [comment]
"visible" check for block esp(check to see if that block is touching air)

**Reply by Zergatul** [reply]
1. This is not easy to implement since block scanning works in separate thread
2. Server plugins that create fake air blocks have been present for more than year already

**Comment by Andy718811** [comment]
Is it possible to add some "time tweaker" function that can trick the server's RPG skill cool time or the day time of the server, so you can sell some items without time limition?

**Reply by Zergatul** [reply]
1. Vanilla packets never send client time. Server always uses its own time
2. I don't know what your RPG mods developers are implemented, but I don't believe they can be this dumb to send to rely on time send from client
3. The only hope is your mod developers forgot to add time check on the server. In this case you have to trigger disabled action on the client and hope server will be ok with it. But cheatutils will not help here, because all these mods are too custom.

**Reply by Andy718811** [reply]
Well, the only thing I know is that they use the pulgin called UltimateShop.
https://www.spigotmc.org/resources/ultimateshop-menus-rewards-shop-daily-shop-in-1-plugin-1-14-1-21-5.110601/

**Reply by Zergatul** [reply]
These inventory menus always have server check, because your client always sends clicks to the server

**Reply by Andy718811** [reply]
Okay, then I wonder if there is a way to "Find the nearest specific mob in a distance and then kill then with killaura"?

**Reply by Zergatul** [reply]
`game.entities.findClosestEntityById(String id)`

**Reply by Andy718811** [reply]
I wrote a code like this and tried it in the Keybinds script, but it failed, the game just crashed.
I just want to know if this will return the mob coordinates, and if I can use this information to automaticallywalk to that place.

```
if (game.entities.getCountById("net.minecraft.world.entity.monster.Stray")>0){
    int test=game.entities.findClosestEntityById("net.minecraft.world.entity.monster.Stray");
    debug.write("test:"+test);
}
```

**Reply by Zergatul** [reply]
"net.minecraft.world.entity.monster.Stray" - this is class name, not id
id is like "minecraft:cow"

**Reply by Andy718811** [reply]
Thank you, but I still how to use the number it returns.
Like this
```
[20:55:58.545689] test:12711
[20:56:02.363766] test:12711
[20:56:29.166956] test:12711
[20:56:53.371229] test:13207
[20:56:53.570721] test:13207
```

**Reply by Zergatul** [reply]
This is integer entity id (not id of entity type, like "minecraft:cow"). You can now attack this entity with `player.attack` if entity is close enough. Or you can tell kill aura to attack only this entity (by saving entity id in global variable, and using kill aura script)

**Reply by Andy718811** [reply]
Well I thought it could find a path and walk toward the mobs and then kill them.
So I still need to walk to the mobs manually?

**Reply by Zergatul** [reply]
Bro, baritone mod does pathing. The single dev who is developing it for years still can't make it work perfectly. You think I have pathing solution in cheatutils?
Baritone can accept commands from chat, if you really want you can tell baritone where to go.

**Reply by someRandomDude-a** [reply]
this is the way, use the chat module to talk to baritone and tell it where to go, baritone will even update pathing in realtime if where you want to go changes, just dont spam baritone too much or you will end up not moving

**Comment by wibi9424** [comment]
https://github.com/user-attachments/assets/39eaf4f0-64cb-4e13-9014-916ca7363b67

3.10.0


this is the reason why im still in 1.21.1 using 3.0.1

so did you manipulate graphics, last time in 1.21.1 make entire screen filled with texture, now some entity got new wierd texture or just transparent,

when im using 1.21.8 in new version is make arrow not visible, but still can see using ESP, and squid has fish texture, i downgrade the version i can see arrow but in different angle got other texture, squid no longer has other texture, but sometimes invisible and in different angle got other texture too

to day im using 1.21.8, im not using pc again because they super lag after 1.12.2 version, i think i cant use this mod again bye

**Reply by Zergatul** [reply]
Are you playing from android? So from what I see, entities sometime get texture of another entity? And if they get invisible part of the texture, entity may become invisible?

**Comment by someRandomDude-a** [comment]
Is there any way I can use the player.target API to find the closest entity to the ray of the crosshair?
Right now I have done a hacky solution by computing a ray and checking perpendicular distance from the player camera ray to the coordinates of all entities on a hotkey, but this code makes the entire game freeze for a few seconds while running the hotkey.

making it set a variable and schedule the task in the events code with an async function does not freeze anymore but still takes about 4~5 seconds to find the closest entity


Problem is, the further the distance you check with the ray method, exponentially higher computations are needed. 


The other method that is really simple is to use `player.target.hasEntity()` and `getEntityID()` , filter for player type and go from there.
The problem here is that the Range is only 3.5 blocks and very limited

**Reply by Zergatul** [reply]
It will be better if I implement this by myself and add a new method. You want to know the entity, which requires a minimum angle to rotate your character, no matter how far away entity is?
`player.target.hasEntity()` I can also add overload for these methods and allow to specify range

**Reply by someRandomDude-a** [reply]
I think that overloading the method will be better so you can specify both range and the cone angle to check.

Another thing I am trying to do is detect if a player is holding a shield or not and the state of that shield (held up or not) so you can selectively disable the shield automatically.
Right now, I just check if the Off-Hand item of the entity is a shield, if so I set a global variable for 100 ticks and count down on the end of every tick so I can kind of track the state. It is preferable to just directly detect the player having a shield held or not

edit: I should specify, I want to detect if the shield is being actively used , not just its existence 

This is how I am doing it right now

```
async void breachSwap(){
    let targetID = variables.getInteger("aimTarget");
    
    if(player.getInteractionDistanceTo(targetID)> 3.5)return;
    let axe = -1;
    let sword = -1;
    let mace = -1; 
    for(int i = 1; i < 10; i++){ 
        let item = inventory.getHotbarItem(i); 
        if(item.item.id == "minecraft:netherite_sword" || item.item.id == "minecraft:diamond_sword")
             sword = i; 
        else if(item.hasEnchantment("breach")) mace = i;

        else if(item.item.id == "minecraft:netherite_axe" || item.item.id == "minecraft:diamond_sword")axe = i;
        } 
        if(sword == -1 && axe != -1)sword = axe; //Use Axe if there is no sword but prefer sword always
        if(!(mace == -1 || sword == -1)){
            
            if(axe != -1 && game.entities.getEquippedOffHandItem(targetID).item.id == "minecraft:shield"
            && variables.getInteger("ShieldCooldown") <= 0)
            {
                inventory.setSelectedHotbarSlot(axe);
                player.attack(targetID);
                variables.setInteger("ShieldCooldown",100);
                ui.systemMessage("Shield cracked");

            }
            inventory.setSelectedHotbarSlot(mace);
            player.attack(targetID); 
            inventory.setSelectedHotbarSlot(sword); 
        }

    
}




async void maceTrackLoop(){
    if(!keys.attack.isDown())variables.setBoolean("Maced",true);// Resets the Attack state if left click is released, allowing for the next strike
    //Critical to stop the method from spamming the target

    let targetId = variables.getInteger("aimTarget");

    if (targetId <= 0 || !game.entities.isAlive(targetId)) {
        return; // no target
    }

    // POSITION DATA
    let px = player.getX();
    let py = player.getY();
    let pz = player.getZ();

    let tx = game.entities.getX(targetId);
    let ty = game.entities.getY(targetId);
    let tz = game.entities.getZ(targetId);

    //DISTANCE
    let dx = tx - px;
    let dy = ty - py;
    let dz = tz - pz;

    let dist = math.sqrt(dx*dx + dz*dz);
    let verticalDiff = py - ty;  // how far WE can fall relative to target
    
    //DISPLAY HUD INFORMATION
    ui.overlayMessage("§aTarget:§f " + game.entities.getName(targetId)
        + "  §aDist:§f " + dist.toStandardString(2)
        + "  §aVerticalDiff:§f " + verticalDiff.toStandardString(2)
    );
    
    //AUTO HIT LOGIC
    let hitRange = player.getInteractionDistanceTo(targetId);
    if (hitRange <= 3.5 
    //&& 
    //inventory.getMainHand().item.id == "minecraft:mace"
    //Checked in the breachSwap function already so is redundant here, Its actually benificial to be holding
    //An Axe or Sword while falling instead
    ) {
        if(variables.getBoolean("Maced") && keys.attack.isDown()){

            // Aim at target head
            player.lookAt(tx, ty + 1.6, tz);

            if(player.target.hasBlock()){//Stops the player from attacking illegaly through blocks
                ui.overlayMessage("Target obstructed by blocks!");
                return;
                }


            // Attack
            breachSwap();
            variables.setBoolean("Maced",false);
        }
    }

}

events.onHandleKeys(() => {
    maceTrackLoop();
});

events.onTickEnd(() => {//Countdown the shield state for 5 seconds (20*5= 100)
    let x = variables.getInteger("ShieldCooldown");
    if(x > 0){
        variables.setInteger("ShieldCooldown",x-1);
    }
});
```

**Reply by Zergatul** [reply]
Someone else asked me about shield state about week ago or something. I will add this too

I think your first script can attack twice

**Reply by someRandomDude-a** [reply]
It only attacks twice if the shield is present and the break timer has expired.
The goal of that script is to use breach swap on a sword or axe to kill extremely fast, We are talking 2 to 3 hits.
Roughly 5 - 6 seconds total. It also disables a shield if it comes up so if you are using a density mace, you can use a wind charge to kill a player in a single sword hit.

All of this assumes you are using strength 2... and if not, just double the time-to-kill

**Reply by Zergatul** [reply]
I don't understand these intricate details... but if this is ok, then ok

**Reply by someRandomDude-a** [reply]
> I think your first script can attack twice

This is intended behavior. It is the relevant part where I need to know if a shield is held up or not

**Reply by someRandomDude-a** [reply]
> Someone else asked me about shield state about week ago or something. I will add this too
> 
> I think your first script can attack twice

ETA on having a full release with these updates? Ideally a 1.21.7-8 build as well if possible.

**Reply by Zergatul** [reply]
I have unfinished big changes for schematica in dev branch. Hopefully will finish them tomorrow. And then release everything else.

**Reply by someRandomDude-a** [reply]
> I have unfinished big changes for schematica in dev branch. Hopefully will finish them tomorrow. And then release everything else.

Sounds great, I am looking through the issues right now to see if there is anything I can fix that has not already been answered

**Reply by Zergatul** [reply]
You can join our matrix, and chat there...

**Reply by someRandomDude-a** [reply]
Hey, can you add a method in the inputs API or the keys API to allow checking the state of any key ? 
Specifically, You can bind to a hotkey.
Ideally:
`keys.isDown("Name of hotkey")` and so on

**Reply by someRandomDude-a** [reply]
> Hey, can you add a method in the inputs API or the keys API to allow checking the state of any key ? Specifically, You can bind to a hotkey. Ideally: `keys.isDown("Name of hotkey")` and so on


Never mind, just read the source code.
This is already implemented.

**Reply by Zergatul** [reply]
The only problem it doesn't work with mouse keys, and this was hard to implement

**Reply by someRandomDude-a** [reply]
This works great.
There are a few things that would greatly improve the usability: setting the player camera rotation using a 3dvector direction.

I am working on a script that raycasts from player eye position to a target entity (stored as ID)
The goal is to find valid positions that can be attacked without intersection with blocks or another entity.

I noticed the entity.getInteractonRange() method directly returns the sqrt of the distance from the bounding box.
Can this method be overloaded to accept a Boolean such that it will calculate the closest valid distance and the required angle from a given position?

Ie: 
`entity.getInteractionRange(position3d origin, position3d max_distance)` returns the direction of the closest valid raycast and it's distance.

**Reply by someRandomDude-a** [reply]
Is there a reason to not use the latest version of loom for new versions of the game? 
And using the latest gradlew release?



> This works great. There are a few things that would greatly improve the usability: setting the player camera rotation using a 3dvector direction.
> 
> I am working on a script that raycasts from player eye position to a target entity (stored as ID) The goal is to find valid positions that can be attacked without intersection with blocks or another entity.
> 
> I noticed the entity.getInteractonRange() method directly returns the sqrt of the distance from the bounding box. Can this method be overloaded to accept a Boolean such that it will calculate the closest valid distance and the required angle from a given position?
> 
> Ie: `entity.getInteractionRange(position3d origin, position3d max_distance)` returns the direction of the closest valid raycast and it's distance.

I am working on implementing it myself

**Reply by Zergatul** [reply]
> entity.getInteractionRange(position3d origin, position3d max_distance) returns the direction of the closest valid raycast and it's distance.

Such method should be named differently.

> Is there a reason to not use the latest version of loom for new versions of the game?

Latest loom caused some problem with my annotation processor for my mixin plugin. This processor will not be required once Mojang removes obfuscation. After this I will upgrade loom.

> And using the latest gradlew release?

It works - I don't touch it

> I am working on implementing it myself

If you want to submit some AI slop, and expect me to spend my time to review it and test it - I don't need this. If you want to submit merge request - you have to make sure code compiles, it works as expected (forge/fabric/neoforge), and it follows the same coding style.

**Reply by someRandomDude-a** [reply]
> If you want to submit some AI slop, and expect me to spend my time to review it and test it - I don't need this. If you want to submit merge request - you have to make sure code compiles, it works as expected (forge/fabric/neoforge), and it follows the same coding style.

Understood, I will make sure it compiles and include example code to showcase the usage

**Comment by someRandomDude-a** [comment]
Suggestion:
Allow the user to fetch waypoints from the player locator bar in later versions, ideally with as high float precision as possible.

This would make it practical to locate the coordinates of players using the locator bar but much more accurately.

Not sure if the waypoints are stored as coordinates or a direction + magnitude etc but being able to even just track the direction vector accurately lets us locate players


I have an example using python for reference on how this can be implemented to track player coordinates 



Intended usage: press F3+C from 2 different locations with the player marker of interest in the center of the locator bar.
The further apart the 2 points are, the better the accuracy of result.


`Intersection.py`

```python
import math
import re

def parse_ray(cmd):
    nums = list(map(float, re.findall(r"-?\d+\.?\d*", cmd)))
    x, y, z, yaw, pitch = nums[0], nums[1], nums[2], nums[3], nums[4]

    yaw_rad = math.radians(yaw)

    dx = -math.sin(yaw_rad)
    dz =  math.cos(yaw_rad)

    return (x, z, dx, dz)

def intersect_rays(ray1, ray2):
    x1, z1, dx1, dz1 = ray1
    x2, z2, dx2, dz2 = ray2

    denom = dx1 * dz2 - dz1 * dx2
    if abs(denom) < 1e-9:
        return None   # Rays are parallel

    t = ((x2 - x1) * dz2 - (z2 - z1) * dx2) / denom

    # Intersection point
    ix = x1 + t * dx1
    iz = z1 + t * dz1
    return ix, iz


ray1 = parse_ray(input("Enter first datapoint:"))
ray2 = parse_ray(input("Enter second datapoint:"))

result = intersect_rays(ray1, ray2)
print("Intersection X,Z =", result)

```

**Comment by someRandomDude-a** [comment]
Add a new event `onInventoryUpdate` 
This would make it possible to only update variables when the inventory changes.

**Comment by Svetozar3211** [comment]
Add a "Control Player" keybind like in this mod https://modrinth.com/mod/freecam#:~:text=Unbound-,Control%20Player,-Transfers%20control%20back

**Reply by Zergatul** [reply]
`freeCam.toggleCameraLock();`

**Comment by csj1818** [comment]


**Reply by Zergatul** [reply]
who told you this is possible

**Reply by csj1818** [reply]
no one i just thought it would be cool to add it, The chat said suggestions so i thought like to suggest like a mod idea

**Reply by csj1818** [reply]
nevermind sry

**Comment by csj1818** [comment]
Sry for bugging you again, This might be a stupid question but in the Key Bindings Youtube Video how did he make the text yellow and search that word?

**Reply by Zergatul** [reply]
Ctrl+F in browser

**Reply by csj1818** [reply]
thanks!

**Comment by csj1818** [comment]
Zergatul Im actully so sorry i keep bugging you its just like im practicly blind and im kinda stupid with this coding stuff and im sorry for bugging you all day, I hope you understand, But whats the key Binding code for Reach under Hacks? I cant find it In refrence

**Reply by Zergatul** [reply]
Not every module was added to API. Reach is one of them. You can better use matrix chat if you have a lot of small questions

**Reply by csj1818** [reply]
ok ill do that in a minnit, but then how do i enable it, just go the Reach Page, Enable it and i have it?

**Reply by Zergatul** [reply]
yes
